Hot-Rod-Transactions-NonXA.asciidoc

```diff
@@ -0,0 +1,311 @@
+Lightweight transactions for Hot Rod
+====================================
+
+*JIRA*: https://issues.jboss.org/browse/ISPN-7807[ISPN-7807]
+
+== Introduction
+
+This document is a design document for lightweight Hot Rod transaction and Synchronization enlistment. Full XA support will be described in another document.
+
+The transaction state is kept in the client during the transaction execution and the server is only aware of the transaction when the client wants to commit.
+
+== Changes to existing code
+
+Since a default `TransactionManager` is needed for the client, a base implementation is extracted from the `EmbeddedTransactionManager` to commons package.
+The `Transaction`'s implementation (and it dependencies) will be moved to commons package.
+
+== Limitation and Requirements
+
+* In the client, the transaction will be enlisted as `Synchronization`.
+XA support will be added later in another document.
+
+* The server will use `REPEATABLE_READ` isolation.
+
+* The server will use the `EmbeddedTransactionManager`.
+It needs to invoke the prepare and commit/rollback phase individually.
+
+* Recovery needs to be enabled in the server.
+It prevents the transaction to be automatically rolled-back when the originated leaves the cluster.
+
+* Any locking is supported.
+
+== Transaction Enlistment
+
+The remote cache is registered in the `TransactionManager` as `Synchronization`.
+
+A global `SynchronizationAdapter` will be implemented.
+It main goal is to register itself and it will map a transaction to the remote caches involved.
+
+If multiple remote caches are involved in a single transaction, it commit will happen orderly between the remote caches.
```

Comment:
You're right that simply running the prepare in parallel would lead to deadlocks. If we want to avoid the roundtrip for each cache, we have to lock keys in order according to <cache-name, key>, and that would mean having a multi-cache prepare command (on the wire, not in the interceptor stack) - and that would be a huge change.

However, on the level of HotRod, you could send the modifications for all caches in one go and let the server handle the details (if it's going to do prepare sequentially, as in the initial implementation, or if it adopts smarter strategy). Even with a sequential impl on server you'll spare some roundtrips to the server.